### DISCUSSION TITLE
Hue User impersonation for Presto doesn't work

### DISCUSSION THREAD

**Discussion by DRavikanth** [cloudera/hue#2449] [Questions]
Is the issue already present in https://github.com/cloudera/hue/issues or discussed in the forum https://discourse.gethue.com? No

Describe the bug: After enabling Hue to impersonate user in Presto, observing 

`(pyhive.exc.OperationalError) Unexpected status code 401 Access Denied: Invalid credentials [SQL: select count(*) from event_detail] (Background on this error at: http://sqlalche.me/e/e3q8)`

Steps to reproduce it?
options configured for SQLAlchemy in hue.ini

 options='{"url": "presto://user:password@localhost:8443/hive/dwh","has_impersonation": true}'

Enabled impersonation in Presto

Hue version or source? (e.g. open source 4.5, CDH 5.16, CDP 1.0...). System info (e.g. OS, Browser...). Hue-4.8.0, open source Docker image. Safari on Mac OS Mojave

**Comment by romainr** [comment]
`invalid credentials`, please specify them https://docs.gethue.com/administrator/configuration/connectors/#presto

**Comment by DRavikanth** [comment]
Hi @romainr, It looks like my description is malformed. Please check the updated url now. I think it looks like a bug here. When impersonation is enabled, Logged in user is being replaced in the connection url, but the password is the one existing in the connection instead of the logged in user.  I think its worth reopening this and taking a look. Thanks.

**Comment by romainr** [comment]
Impersonation means using `user:password` as the credentials but then performing the operations as the other users, so this sounds correct (the password of the logged in user is unknown here, or should be set toe `${PASSWORD}` in the connection string cf. the docs (and this was not tested in a while, but it used to prompt the user for it then)

**Comment by DRavikanth** [comment]
When we create a Notebook for Presto via JDBC, it will pop up for the authentication. But, with SQLAlchemy, the user will not be prompted for password. I think this is the problem.

**Comment by DRavikanth** [comment]
@romainr, May I know the version of PyHive we are using for Hue?  I tried finding but couldn't locate the version.

**Comment by romainr** [comment]
https://github.com/gethue/PyHive bu will probably rebase on https://github.com/dropbox/PyHive soon

**Comment by byungnam** [comment]
I also faced this error while enabling Trino connector "has_impersonation" option with interface=SqlAlchemy.

Luckily, this can be resolved by [PyHive 0.6.3 principal_username option](https://github.com/dropbox/PyHive/blob/b21c507a24ed2f2b0cf15b0b6abb1c43f31d3ee0/pyhive/presto.py#L136-L151 ).

But then the problem now is, we cannot send the principal_username dynamically with current implementation. If we set a fixed `principal_username` in hue.ini file, then the whole user will be impersonated to that user. We can add `self.options['connect_args']['principal_username'] = self.user.username` at sql_alchemy.py, but then it will always try to impersonate even if you don't want to.

@romainr I think 'has_impersonation' option needs to be revised to use the above `principal_username` option. At least, we need to fix it as it is not working when password authentication is enabled.

**Reply by sbbagal13** [reply]
@byungnam I am also facing same issue for connecting Hue to Trino , If possible could you please share code snippet from hue.ini where you setup connection details for Trino

**Reply by byungnam** [reply]
@sbbagal13 
Here is my hue.ini. Hope this can help you.
```
[notebook]
  [[interpreters]]
    [[[trino-hadoop-rabbit]]]
      name=Trino-Hadoop-Rabbit
      interface=sqlalchemy
      options='{"url":"trino://USER:PW@hadoop-rabbit-trino.onkakao.net:8443/hadoop_rabbit/default", "connect_args":"{\"protocol\":\"https\"}", "has_impersonation": true}'
```

**Reply by sbbagal13** [reply]
@byungnam Thanks for responding, Somehow when I install only pyHive (0.6.3 or 0.6.5) I am getting trino dialect not found And if I add trino package on top of it , error coming as unrecognize parameter "has_impersonation" .
Also as mentioned in your earlier comment , how did you provided  principal_username in it. I am using **gethue/hue:4.10.0**  docker image and Trino 393 ..  Any suggestion to resolve the issue ?

**Comment by romainr** [comment]
:+1: to make it smarter. AFAIK this is only used for Hive/Presto (and DaskSql):
https://github.com/cloudera/hue/blob/master/desktop/libs/notebook/src/notebook/connectors/sql_alchemy.py#L175

Also just rebased https://github.com/gethue/PyHive on https://github.com/dropbox/PyHive to get the Trino commit.
